File: iosApp/iosApp/Presentation/Components/VinylSleeve.swift
import SwiftUI

struct VinylSleeve: View {
    let albumTitle: String
    let artistName: String
    let year: String
    var side: String = "SIDE A"
    let imageUrl: String?
    let isPlaying: Bool

    var body: some View {
        ZStack {
            // 1. Underglow: 220x220 radial gold gradient, blurred
            RadialGradient(
                colors: [Color.accentColor.opacity(0.25), Color.clear],
                center: .center,
                startRadius: 0,
                endRadius: 110,
            )
            .frame(width: 220, height: 220)
            .blur(radius: 32)

            // Wrapper that applies the 3D perspective transforms
            ZStack(alignment: .leading) {
                // 2. LP record disc: 248x248, offset to the right when playing
                Canvas { context, size in
                    let rect = CGRect(origin: .zero, size: size)
                    let center = CGPoint(x: size.width / 2, y: size.height / 2)
                    let outerRadius = size.width / 2

                    // Draw the black matte vinyl disc body
                    context.fill(Path(ellipseIn: rect), with: .color(Color(hex: 0x0F0F11)))

                    // Draw concentric vinyl grooves (concentric circles)
                    var grooveRadius = outerRadius - 12 // 12pt
                    let minGrooveRadius: CGFloat = 42
                    let grooveSpacing: CGFloat = 6

                    while grooveRadius > minGrooveRadius {
                        let grooveRect = CGRect(
                            x: center.x - grooveRadius,
                            y: center.y - grooveRadius,
                            width: grooveRadius * 2,
                            height: grooveRadius * 2,
                        )
                        context.stroke(
                            Path(ellipseIn: grooveRect),
                            with: .color(Color(hex: 0x1E1E21)),
                            style: StrokeStyle(lineWidth: 1),
                        )
                        grooveRadius -= grooveSpacing
                    }

                    // Draw the center 72pt gold label disc (radius = 36pt)
                    let labelRect = CGRect(
                        x: center.x - 36,
                        y: center.y - 36,
                        width: 72,
                        height: 72,
                    )
                    context.fill(Path(ellipseIn: labelRect), with: .color(.accentColor))

                    // Draw center hole 6pt (radius = 3pt) in deepest background color bg0
                    let holeRect = CGRect(
                        x: center.x - 3,
                        y: center.y - 3,
                        width: 6,
                        height: 6,
                    )
                    context.fill(Path(ellipseIn: holeRect), with: .color(.bg0))
                }
                .frame(width: 248, height: 248)
                .offset(x: isPlaying ? 64 : 0)
                .animation(.easeInOut(duration: 0.5), value: isPlaying)

                // 3. Sleeve: 256x256, 4pt radius, dark gradient + repeating 135° stripe
                ZStack {
                    // Background artwork or gradient
                    if let imageUrl, let url = URL(string: imageUrl) {
                        JrrAsyncImage(url: url) { image in
                            image
                                .resizable()
                                .aspectRatio(contentMode: .fill)
                        } placeholder: {
                            LinearGradient(
                                colors: [.bg2, .bg0],
                                startPoint: .top,
                                endPoint: .bottom,
                            )
                        }
                    } else {
                        LinearGradient(
                            colors: [.bg2, .bg0],
                            startPoint: .top,
                            endPoint: .bottom,
                        )
                    }

                    // Sleeve decoration (stripes & inset border)
                    Canvas { context, size in
                        // Draw 135-degree repeating gold stripe at 6% opacity
                        let step: CGFloat = 20
                        let lineCount = Int((size.width + size.height) / step)
                        for i in 0 ... lineCount {
                            let offset = CGFloat(i) * step
                            var path = Path()
                            path.move(to: CGPoint(x: offset, y: 0))
                            path.addLine(to: CGPoint(x: offset - size.height, y: size.height))
                            context.stroke(
                                path,
                                with: .color(Color.accentColor.opacity(0.06)),
                                style: StrokeStyle(lineWidth: 2),
                            )
                        }

                        // 1-px inset border at 18px (which is 6pt)
                        let inset: CGFloat = 6
                        let insetRect = CGRect(
                            x: inset,
                            y: inset,
                            width: size.width - 2 * inset,
                            height: size.height - 2 * inset,
                        )
                        context.stroke(
                            Path(insetRect),
                            with: .color(.accentSoft),
                            style: StrokeStyle(lineWidth: 1),
                        )
                    }

                    // Overlaid text content inside the sleeve
                    VStack(alignment: .leading) {
                        // Top-Left metadata: "SIDE A · YEAR"
                        HStack {
                            Text("\(side) · \(year)".uppercased())
                                .font(AppFont.ibmPlexMono(size: 11, weight: .regular))
                                .tracking(0.2)
                                .foregroundColor(.accentColor)
                                .monospacedDigit()
                            Spacer()
                        }

                        Spacer()

                        // Bottom-Left Titles
                        VStack(alignment: .leading, spacing: 4) {
                            Text(albumTitle.uppercased())
                                .font(AppFont.inter(size: 22, weight: .bold))
                                .tracking(-0.5)
                                .foregroundColor(.textPrimary)
                                .lineLimit(2)

                            Text(artistName.uppercased())
                                .font(AppFont.ibmPlexMono(size: 11, weight: .regular))
                                .tracking(1.6)
                                .foregroundColor(.textSecondary)
                                .lineLimit(1)
                        }
                    }
                    .padding(18)
                }
                .frame(width: 256, height: 256)
                .background(Color.bg3) // Fallback backplate
                .cornerRadius(4)
            }
            .frame(width: 256, height: 256)
            .rotation3DEffect(.degrees(8), axis: (x: 1, y: 0, z: 0))
            .rotation3DEffect(.degrees(-14), axis: (x: 0, y: 1, z: 0))
            .rotation3DEffect(.degrees(-3), axis: (x: 0, y: 0, z: 1))
        }
        .frame(width: 260, height: 260)
    }
}

#Preview {
    ZStack {
        Color.bg0.ignoresSafeArea()
        VinylSleeve(
            albumTitle: "Dark Side of the Moon",
            artistName: "Pink Floyd",
            year: "1973",
            imageUrl: nil,
            isPlaying: true,
        )
    }
}

File: composeUi/src/commonMain/kotlin/com/jrr/jrrkmp_native_ui/presentation/components/VinylSleeve.kt
package com.jrr.jrrkmp_native_ui.presentation.components

import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.blur
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import coil3.compose.AsyncImage
import com.jrr.jrrkmp_native_ui.core.theme.AppColors
import com.jrr.jrrkmp_native_ui.core.theme.AppTypography

@Composable
fun VinylSleeve(
    albumTitle: String,
    artistName: String,
    year: String,
    side: String = "SIDE A",
    imageUrl: String?,
    isPlaying: Boolean,
    modifier: Modifier = Modifier
) {
    // Outer size 260dp x 260dp
    Box(
        modifier = modifier
            .size(260.dp),
        contentAlignment = Alignment.Center
    ) {
        // 1. Underglow: 220dp x 220dp radial gold gradient, blurred
        Box(
            modifier = Modifier
                .size(220.dp)
                .blur(32.dp)
                .background(
                    brush = Brush.radialGradient(
                        colors = listOf(
                            AppColors.accent.copy(alpha = 0.25f),
                            Color.Transparent
                        )
                    ),
                    shape = RoundedCornerShape(110.dp)
                )
        )

        // Wrapper Box that applies the 3D perspective transforms
        // transform: rotateX(8) rotateY(-14) rotateZ(-3)
        Box(
            modifier = Modifier
                .fillMaxSize()
                .graphicsLayer {
                    rotationX = 8f
                    rotationY = -14f
                    rotationZ = -3f
                    cameraDistance = 12f * density
                },
            contentAlignment = Alignment.CenterStart
        ) {
            // 2. LP record disc: 248dp x 248dp, offset 64dp right of sleeve when playing
            val discOffset by animateDpAsState(
                targetValue = if (isPlaying) 64.dp else 0.dp,
                animationSpec = tween(durationMillis = 500),
                label = "DiscOffsetAnimation"
            )

            Box(
                modifier = Modifier
                    .size(248.dp)
                    .offset(x = discOffset)
                    .align(Alignment.CenterStart)
            ) {
                Canvas(modifier = Modifier.fillMaxSize()) {
                    val canvasWidth = size.width
                    val canvasHeight = size.height
                    val center = Offset(x = canvasWidth / 2f, y = canvasHeight / 2f)
                    val outerRadius = canvasWidth / 2f

                    // Draw the black matte vinyl disc body
                    drawCircle(
                        color = Color(0xFF0F0F11),
                        radius = outerRadius
                    )

                    // Draw concentric vinyl grooves (concentric circles)
                    var grooveRadius = outerRadius - 12.dp.toPx()
                    val minGrooveRadius = 42.dp.toPx()
                    val grooveSpacing = 6.dp.toPx()

                    while (grooveRadius > minGrooveRadius) {
                        drawCircle(
                            color = Color(0xFF1E1E21),
                            radius = grooveRadius,
                            style = Stroke(width = 1.dp.toPx())
                        )
                        grooveRadius -= grooveSpacing
                    }

                    // Draw the center 72dp gold label disc (diameter = 72dp -> radius = 36dp)
                    drawCircle(
                        color = AppColors.accent,
                        radius = 36.dp.toPx()
                    )

                    // Draw center hole 6dp (radius = 3dp) in deepest background color bg0
                    drawCircle(
                        color = AppColors.bg0,
                        radius = 3.dp.toPx()
                    )
                }
            }

            // 3. Sleeve: 256dp x 256dp, 4dp radius, dark gradient + repeating 135° gold stripe
            Box(
                modifier = Modifier
                    .size(256.dp)
                    .clip(RoundedCornerShape(4.dp))
                    .background(
                        brush = Brush.verticalGradient(
                            colors = listOf(AppColors.bg2, AppColors.bg0)
                        )
                    )
            ) {
                // Background album artwork if available
                if (!imageUrl.isNullOrEmpty()) {
                    val resolver = com.jrr.jrrkmp_native_ui.presentation.LocalArtworkResolver.current
                    val finalImageUrl = androidx.compose.runtime.remember(imageUrl, resolver) {
                        resolver.resolve(imageUrl)
                    }
                    AsyncImage(
                        model = finalImageUrl,
                        contentDescription = "Album Artwork",
                        modifier = Modifier.fillMaxSize(),
                        contentScale = ContentScale.Crop
                    )
                }

                // Sleeve decoration (stripes & inset border)
                Canvas(modifier = Modifier.fillMaxSize()) {
                    // Draw 135-degree repeating gold stripe at 6% opacity
                    val step = 20.dp.toPx()
                    val lineCount = (size.width + size.height) / step
                    for (i in 0..lineCount.toInt()) {
                        val offset = i * step
                        drawLine(
                            color = AppColors.accent.copy(alpha = 0.06f),
                            start = Offset(x = offset, y = 0f),
                            end = Offset(x = offset - size.height, y = size.height),
                            strokeWidth = 2.dp.toPx()
                        )
                    }

                    // 1-px inset border at 18px (which is 6dp)
                    val inset = 6.dp.toPx()
                    drawRect(
                        color = AppColors.accentSoft,
                        topLeft = Offset(x = inset, y = inset),
                        size = Size(
                            width = size.width - 2 * inset,
                            height = size.height - 2 * inset
                        ),
                        style = Stroke(width = 1.dp.toPx())
                    )
                }

                // Overlaid text content inside the sleeve
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(18.dp) // Sits exactly inside the inset border line
                ) {
                    // Top-Left metadata: "SIDE A · YEAR"
                    Text(
                        text = "$side · $year".uppercase(),
                        style = AppTypography.monoLabel.copy(
                            color = AppColors.accent
                        ),
                        modifier = Modifier.align(Alignment.TopStart)
                    )

                    // Bottom-Left Titles
                    Column(
                        modifier = Modifier.align(Alignment.BottomStart)
                    ) {
                        Text(
                            text = albumTitle.uppercase(),
                            style = AppTypography.screenTitle,
                            maxLines = 2
                        )
                        Text(
                            text = artistName.uppercase(),
                            style = AppTypography.sectionHeading.copy(
                                color = AppColors.text2
                            ),
                            maxLines = 1,
                            modifier = Modifier.padding(top = 4.dp)
                        )
                    }
                }
            }
        }
    }
}
